File: typescript/src/domain/usecases/SubmitPostUseCase.ts
import Post, { newPost } from "../entities/Post";
import PostRepository from "../repositories/PostRepository";
import UserRepository from "../repositories/UserRepository";
import UserNotFoundError from "./errors/UserNotFoundError";

export default class SubmitPostUseCase {
  private postRepository: PostRepository;
  private userRepository: UserRepository;

  constructor(postRepository: PostRepository, userRepository: UserRepository) {
    this.postRepository = postRepository
    this.userRepository = userRepository
  }

  run(userId: string, postText: string): Post {
    const registeredUser = this.userRepository.getUserById(userId)
    if (!registeredUser)
      throw new UserNotFoundError(userId)

    if(postText.toLocaleLowerCase().includes("orange"))
      throw new InappropriateLanguageError(postText)

    const postToStore: Post = newPost(postText, userId)
    this.postRepository.store(postToStore)
    return postToStore
  }

}

export class InappropriateLanguageError extends Error {
  constructor(postText: string) {
    super("Post text contains inappropriate language: " + postText)
    Object.setPrototypeOf(this, InappropriateLanguageError.prototype);
  }
}

File: java/src/main/java/tech/qmates/openchat/domain/usecase/SubmitPostUseCase.java
package tech.qmates.openchat.domain.usecase;

import tech.qmates.openchat.domain.UTCClock;
import tech.qmates.openchat.domain.UserNotFoundException;
import tech.qmates.openchat.domain.entity.Post;
import tech.qmates.openchat.domain.entity.RegisteredUser;
import tech.qmates.openchat.domain.repository.PostRepository;
import tech.qmates.openchat.domain.repository.UserRepository;

import java.util.List;
import java.util.UUID;

public class SubmitPostUseCase {

    public static final List<String> FORBIDDEN_WORDS = List.of("orange", "elephant");

    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final UTCClock clock;

    public SubmitPostUseCase(PostRepository postRepository, UserRepository userRepository, UTCClock clock) {
        this.postRepository = postRepository;
        this.userRepository = userRepository;
        this.clock = clock;
    }

    public Post run(UUID authorUserId, String postText) throws UserNotFoundException, InappropriateLanguageException {
        RegisteredUser user = userRepository.getUserById(authorUserId);
        if (user == null)
            throw new UserNotFoundException(authorUserId);

        if(containsForbiddenWords(postText))
            throw new InappropriateLanguageException(postText);

        Post postToStore = new Post(
            UUID.randomUUID(),
            authorUserId,
            postText,
            clock.now()
        );
        postRepository.store(postToStore);
        return postToStore;
    }

    private static boolean containsForbiddenWords(String postText) {
        return FORBIDDEN_WORDS.stream()
            .anyMatch((word) -> postText.toLowerCase().contains(word));
    }

    public static class InappropriateLanguageException extends Exception {
        public InappropriateLanguageException(String postText) {
            super("Post text contains inappropriate language: " + postText);
        }
    }
}
